File: packages/android/src/main/java/com/kozmos/components/Alert/Alert.kt
package com.kozmos.components.alert

import com.kozmos.tokens.KozmosDimensions

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosThemeTokens

enum class AlertStatus {
    Default, Info, Success, Warning, Error
}

@Composable
fun KozmosAlert(
    modifier: Modifier = Modifier,
    status: AlertStatus = AlertStatus.Default,
    content: @Composable androidx.compose.foundation.layout.RowScope.() -> Unit
) {
    val borderColor = alertBorderColor(status)

    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(KozmosThemeTokens.semanticsSurface0, RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
            .border(1.dp, borderColor, RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
            .padding(KozmosDimensions.primitivesLayoutSpacing200),
        verticalAlignment = Alignment.Top
    ) {
        content()
    }
}

@Composable
fun KozmosAlertIcon(
    status: AlertStatus = AlertStatus.Default,
    modifier: Modifier = Modifier,
    customIcon: ImageVector? = null
) {
    val icon = when (status) {
        AlertStatus.Default -> Icons.Default.Info
        AlertStatus.Info -> Icons.Default.Info
        AlertStatus.Success -> Icons.Default.CheckCircle
        AlertStatus.Warning -> Icons.Default.Warning
        AlertStatus.Error -> Icons.Default.Warning
    }
    Icon(
        imageVector = customIcon ?: icon,
        contentDescription = null,
        tint = alertForegroundColor(status),
        modifier = modifier.padding(end = KozmosDimensions.primitivesLayoutSpacing150)
    )
}

@Composable
fun KozmosAlertContent(
    modifier: Modifier = Modifier,
    content: @Composable androidx.compose.foundation.layout.ColumnScope.() -> Unit
) {
    Column(modifier = modifier) {
        content()
    }
}

@Composable
fun KozmosAlertTitle(
    title: String,
    modifier: Modifier = Modifier,
    color: Color = KozmosThemeTokens.primitivesColorsForeground100
) {
    Text(
        text = title,
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        color = color,
        modifier = modifier
    )
}

@Composable
fun KozmosAlertDescription(
    description: String,
    modifier: Modifier = Modifier,
    color: Color = KozmosThemeTokens.primitivesColorsForeground500
) {
    Text(
        text = description,
        style = MaterialTheme.typography.bodyMedium,
        color = color,
        modifier = modifier
    )
}

@Composable
@ReadOnlyComposable
internal fun alertForegroundColor(status: AlertStatus): Color = when (status) {
    AlertStatus.Default -> KozmosThemeTokens.primitivesColorsForeground0
    AlertStatus.Info -> KozmosThemeTokens.primitivesColorsTheme600
    AlertStatus.Success -> KozmosThemeTokens.primitivesColorsEmotionalSuccess900
    AlertStatus.Warning -> KozmosThemeTokens.primitivesColorsEmotionalAlert900
    AlertStatus.Error -> KozmosThemeTokens.primitivesColorsEmotionalDanger600
}

@Composable
@ReadOnlyComposable
private fun alertBorderColor(status: AlertStatus): Color = when (status) {
    AlertStatus.Default -> KozmosThemeTokens.primitivesColorsForeground500
    else -> alertForegroundColor(status)
}

File: packages/ios/Sources/Components/Alert/Alert.swift
import SwiftUI

public struct KozmosAlert<Content: View>: View {
    let variant: KozmosAlertVariant
    let content: Content

    public enum KozmosAlertVariant {
        case `default`, info, success, warning, destructive

        var foregroundColor: Color {
            switch self {
            case .default: return KozmosColors.primitivesColorsForeground0
            case .info: return KozmosColors.primitivesColorsTheme600
            case .success: return KozmosColors.primitivesColorsEmotionalSuccess900
            case .warning: return KozmosColors.primitivesColorsEmotionalAlert900
            case .destructive: return KozmosColors.primitivesColorsEmotionalDanger600
            }
        }

        var borderColor: Color {
            switch self {
            case .default: return KozmosColors.primitivesColorsForeground500
            default: return foregroundColor
            }
        }

        var icon: String {
            switch self {
            case .default: return "info.circle"
            case .info: return "info.circle"
            case .success: return "checkmark.circle"
            case .warning: return "exclamationmark.triangle"
            case .destructive: return "xmark.circle"
            }
        }
    }

    public init(variant: KozmosAlertVariant = .default, @ViewBuilder content: () -> Content) {
        self.variant = variant
        self.content = content()
    }

    public var body: some View {
        HStack(alignment: .top, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
            content
        }
        .padding()
        .background(KozmosColors.semanticsSurface0)
        .cornerRadius(KozmosDimensions.primitivesLayoutRadius100)
        .overlay(
            RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100)
                .stroke(variant.borderColor, lineWidth: 1)
        )
    }
}

public struct KozmosAlertIcon: View {
    let variant: KozmosAlert<AnyView>.KozmosAlertVariant

    public init(variant: KozmosAlert<AnyView>.KozmosAlertVariant = .default) {
        self.variant = variant
    }

    public var body: some View {
        Image(systemName: variant.icon)
            .foregroundColor(variant.foregroundColor)
            .font(.title3)
    }
}

public struct KozmosAlertContent<Content: View>: View {
    let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            content
        }
    }
}

public struct KozmosAlertTitle: View {
    let title: String
    let color: Color

    public init(_ title: String, color: Color = KozmosColors.primitivesColorsForeground0) {
        self.title = title
        self.color = color
    }

    public var body: some View {
        Text(title)
            .font(.headline)
            .bold()
            .foregroundColor(color)
    }
}

public struct KozmosAlertDescription: View {
    let description: String
    let color: Color

    public init(_ description: String, color: Color = KozmosColors.primitivesColorsForeground500) {
        self.description = description
        self.color = color
    }

    public var body: some View {
        Text(description)
            .font(.subheadline)
            .foregroundColor(color)
    }
}
